[SDPA] Replace `masked_fill_` with `aten::where`

sijiac opened the pull request:
Summary:
full context in D59385876

Based on the offline discussion with PT2 folks, we switched to change the SDPA impl to mitigate the AOTI lowering issue

Test Plan: PYTORCH_TEST_FBCODE=1 buck2 run  mode/opt -c=python.package_style=inplace -c fbcode.enable_gpu_sections=true -c fbcode.platform=platform010 -c fbcode.split-dwarf=true caffe2/test/inductor:test_inductor -- -r test_sdpa_inference_mode_aot_compile

Differential Revision: D59495634


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

zou3519 commented on aten/src/ATen/native/transformers/attention.cpp:
```diff
@@ -526,11 +526,9 @@ std::optional<Tensor> convert_boolean_attn_mask(const std::optional<Tensor>& att
-    new_attn_mask.masked_fill_(
-        attn_mask->logical_not(), -std::numeric_limits<double>::infinity());
-    return new_attn_mask;
+    auto new_attn_mask = at::zeros_like(attn_mask.value(), dtype);
+    return at::where(attn_mask->logical_not(), -std::numeric_limits<double>::infinity(), new_attn_mask);
```
Does `at::where(attn_mask->logical_not(),-std::numeric_limits<double>::infinity(), 0.)` work? As written we are allocating two tensors; we should be able to get this to allocate just one tensor.

sijiac commented:
@pytorchbot rebase -b main

peterbell10 commented on aten/src/ATen/native/transformers/attention.cpp:
```diff
@@ -526,11 +527,8 @@ std::optional<Tensor> convert_boolean_attn_mask(const std::optional<Tensor>& att
-    new_attn_mask.masked_fill_(
-        attn_mask->logical_not(), -std::numeric_limits<double>::infinity());
-    return new_attn_mask;
+    return at::where(attn_mask->logical_not(), -std::numeric_limits<double>::infinity(), at::scalar_tensor(0.0, at::TensorOptions().dtype(dtype)));
```
NIT: You can also avoid the `logical_not` by reversing the argument order.
```suggestion
    return at::where(*attn_mask, 0.0, at::scalar_tensor(-std::numeric_limits<double>::infinity(), at::TensorOptions().dtype(dtype)));
```

sijiac commented:
@pytorchbot rebase -b main

justinchuby reviewed (approved):
ONNX lgtm

sijiac commented:
@thiagocrepaldi, did you have idea of the onnx failure? We did a try in the `viable/strict` branch, which failed with the error
```
AssertionError: Expected error message: n=copy_, n.args[0]=zeros_like, placeholders={ NOT in Unknown input type from input: expand
```

justinchuby commented:
Removing https://github.com/pytorch/pytorch/blob/68ad3eb722a429db738f93c73af82f700bc89ef5/test/onnx/test_fx_to_onnx_with_onnxruntime.py#L1271-L1274 may help

justinchuby commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -1263,15 +1263,11 @@ def create_model():
+    @pytorch_test_common.xfail(
         error_message="Could not find an implementation for Trilu(14) node",
         reason="ORT error during op level dubug",
         model_type=pytorch_test_common.TorchModelType.TORCH_NN_MODULE,
```
```suggestion
```

sijiac committed:
[SDPA] Replace `masked_fill_` with `aten::where` (#130281)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/130281

full context in D59385876

Based on the offline discussion with PT2 folks, we switched to change the SDPA impl to mitigate the AOTI lowering issue

Test Plan: PYTORCH_TEST_FBCODE=1 buck2 run  mode/opt -c=python.package_style=inplace -c fbcode.enable_gpu_sections=true -c fbcode.platform=platform010 -c fbcode.split-dwarf=true caffe2/test/inductor:test_inductor -- -r test_sdpa_inference_mode_aot_compile

Reviewed By: zou3519, bdhirsh

Differential Revision: D59495634

sijiac commented:
@pytorchbot merge
